Add a function for running circuits in a shuffled order with readout error benchmarking.

ddddddanni opened the pull request:
This pull request adds a function to run the circuits in a shuffled order with readout error benchmarking.

This pull request introduces a new function, `run_shuffled_with_readout_benchmarking`, to the `cirq.experiments` module. The function allows users to execute a set of quantum circuits in a randomized (shuffled) order while performing readout error benchmarking. 

Open questions:
1. I put the method under experiment folder. The reason is this method performs readout benchmarking, which is a little bit similar to single_qubit_readout_calibration which is also under this folder.
2. I use the qubits extracts from the input circuits to generate random readout benchmarking circuits. Not sure if this is the intended way to do so. 
3. For generating random readout calibration circuits, this is based on the random bit strings (in a straight forward way). Not sure if there are better ways to generate those circuits.

NoureldinYosri commented on cirq-core/cirq/experiments/shuffle_circuits_with_readout_benchmarking.py:
```diff
@@ -0,0 +1,138 @@
+    qubits = sorted(qubits)
+
+    # Generate the readout calibration circuits
+    rng = np.random.default_rng()
```
take a `np.Generator` or a seed as input instead of constructing one inline

ddddddanni replied:
Done. I added this as the method input.

NoureldinYosri commented on cirq-core/cirq/experiments/shuffle_circuits_with_readout_benchmarking.py:
```diff
@@ -0,0 +1,138 @@
+    num_random_bitstrings: int = 100,
+    readout_repetitions: int = 1000,
+    ) -> tuple[np.ndarray, dict["cirq.Qid": tuple[float, float]]]:
+    # tuple[np.ndarray, dict[cirq.Qid: tuple[float, float]]]:
```
please cleanup this part

ddddddanni replied:
Done.

NoureldinYosri commented on cirq-core/cirq/experiments/shuffle_circuits_with_readout_benchmarking.py:
```diff
@@ -0,0 +1,138 @@
+        raise ValueError("Input circuits must be of type cirq.Circuit." )
+    # Check input_circuits have measurements
+    for circuit in input_circuits:
+        if not any(isinstance(op.gate, ops.MeasurementGate) for op in circuit.all_operations()):
```
call `protocols.is_measurement` instead of `isinstance`

ddddddanni replied:
Done.

NoureldinYosri commented on cirq-core/cirq/experiments/shuffle_circuits_with_readout_benchmarking.py:
```diff
@@ -0,0 +1,138 @@
+    rng = np.random.default_rng()
+    x_or_I = lambda bit: ops.X if bit == 1 else ops.I
+
+    random_bitstrings = np.random.randint(0, 2, size=(num_random_bitstrings, len(qubits)), dtype=np.int32)
```
use `rng.integers` instead of `np.random.randint` and generally use `rng.` instead of the `numpy.random` module

ddddddanni replied:
Sure!

NoureldinYosri commented on cirq-core/cirq/experiments/shuffle_circuits_with_readout_benchmarking.py:
```diff
@@ -0,0 +1,138 @@
+    all_circuits = input_circuits + readout_calibration_circuits
+    all_repetitions = circuit_repetitions + [readout_repetitions] * len(readout_calibration_circuits)
+
+    shuf_order = np.arange(len(all_circuits))
```
you can get the order in one call

```py3
shuf_order = rng.permutation(len(all_circuits))
```

ddddddanni replied:
Done.

NoureldinYosri commented on cirq-core/cirq/experiments/shuffle_circuits_with_readout_benchmarking.py:
```diff
@@ -0,0 +1,138 @@
+    rng.shuffle(shuf_order)
+    unshuf_order = np.zeros_like(shuf_order)
+    unshuf_order[shuf_order] = np.arange(len(all_circuits))
+    shuffled_circuits = [all_circuits[_] for _ in shuf_order]
```
use `i` instead `_`

ddddddanni replied:
Done.

NoureldinYosri commented on cirq-core/cirq/experiments/shuffle_circuits_with_readout_benchmarking.py:
```diff
@@ -0,0 +1,138 @@
+    trial_idx = 0
+    for trial_result in unshuffled_readout_measurements:
+        trial_result = trial_result.astype(np.int64)  # Cast to int64
+        sample_counts = np.einsum('ij->j', trial_result)
```
nit: replace `einsum` with `np.sum(..., axis=0)`

ddddddanni replied:
Done.

NoureldinYosri reviewed (changes requested):
please try to breakdown the method into smaller functions

ddddddanni commented:
> please try to breakdown the method into smaller functions

Sure, I breakdown the method into several sub methods.

ddddddanni commented:
> Regarding question 1, I think @dstrain115 was saying that this should go in `cirq.contrib`, but I don't have a strong preference myself. Regarding question 2, maybe we let that be the default behavior and allow the user to optionally specify qubits to use for readout benchmarking. Regadring question 3, that sounds good to me.

Done! I moved the whole stuff to contrib directory, and also added an optional 'qubit' input to the method. Additionally, I also added a test to reflect the change.

NoureldinYosri commented on cirq-core/cirq/contrib/shuffle_circuits/__init__.py:
```diff
@@ -0,0 +1,18 @@
+# Copyright 2019 The Cirq Developers
```
the license year is the year in which the file is first created, please update here an elsewhere
```suggestion
# Copyright 2025 The Cirq Developers
```

ddddddanni replied:
Done

NoureldinYosri commented on cirq-core/cirq/contrib/shuffle_circuits/shuffle_circuits_with_readout_benchmarking.py:
```diff
@@ -0,0 +1,194 @@
+        num_random_bitstrings: The number of random bitstrings for measuring readout.
+        readout_repetitions: The number of repetitions for each readout bitstring.
+        qubits: The qubits to benchmark readout errors. If None, all qubits in the 
+        input_circuits are used.
```
nit:
```suggestion
                input_circuits are used.
```

NoureldinYosri commented on cirq-core/cirq/contrib/shuffle_circuits/shuffle_circuits_with_readout_benchmarking.py:
```diff
@@ -0,0 +1,194 @@
+    input_circuits: list[circuits.Circuit],
+    sampler: work.Sampler,
+    circuit_repetitions: int | list[int],
+    rng: np.random.Generator,
```
allow passing a seed
```py3
    rng_or_seed: Union[np.random.Generator, int] = None,
    
    ...
    
    rng = rng_or_seed if isinstance(rng_or_seed, np.random.Generator) else np.random.default_rng(rng_or_seed)
```

ddddddanni replied:
Done

NoureldinYosri commented on cirq-core/cirq/contrib/shuffle_circuits/shuffle_circuits_with_readout_benchmarking.py:
```diff
@@ -0,0 +1,194 @@
+                                                         list[int], np.ndarray]:
+    """Shuffles the input circuits and readout calibration circuits."""
+    shuf_order = rng.permutation(len(all_circuits))
+    unshuf_order = np.zeros_like(shuf_order)
```
nit:
```suggestion
    unshuf_order = np.zeros_like(shuf_order)    # Inverse permutation
```

NoureldinYosri commented on cirq-core/cirq/contrib/shuffle_circuits/shuffle_circuits_with_readout_benchmarking.py:
```diff
@@ -0,0 +1,194 @@
+    for bitstr in random_bitstrings:
+        readout_calibration_circuits.append(
+            circuits.Circuit(
+                [x_or_i(bit)(qubit) for bit, qubit in zip(bitstr, qubits)]
```
this is a matter of taste, so it is upto you but I think an array is better here

```py3
bit_to_gate = (cirq.I, cirq.X)
```

```suggestion
                [bit_to_gate[bit](qubit) for bit, qubit in zip(bitstr, qubits)]
```

ddddddanni replied:
Done

NoureldinYosri reviewed (changes requested):
from a coding point of view, overall looks good % nits

---

@eliottrosenberg can you check the correctness

eliottrosenberg commented on cirq-core/cirq/contrib/shuffle_circuits/shuffle_circuits_with_readout_benchmarking.py:
```diff
@@ -0,0 +1,194 @@
+
+    # Run the shuffled circuits and measure
+    results = sampler.run_batch(shuffled_circuits, repetitions=all_repetitions)
+    shuffled_measurements = [res[0].measurements["m"] for res in results]
```
This assumes that the input circuits have a single set of measurements with key "m". Maybe it would be better to keep the dictionary `measurements` or even the results object `res[0]` instead of extracting the numpy array `measurements["m"]`.

ddddddanni replied:
Sure. In the new version, I made 
```
results = sampler.run_batch(shuffled_circuits, repetitions=all_repetitions)
shuffled_measurements = [res[0] for res in results]
unshuffled_measurements = [shuffled_measurements[i] for i in unshuf_order]
```
Basically keep the res[0].

eliottrosenberg commented on cirq-core/cirq/contrib/shuffle_circuits/shuffle_circuits_with_readout_benchmarking.py:
```diff
@@ -0,0 +1,194 @@
+    readout_error_rates = _analyze_readout_results(
+        unshuffled_readout_measurements, random_bitstrings, readout_repetitions, qubits)
+
+    return unshuffled_measurements, readout_error_rates
```
Actually, we can return `unshuffled_measurements[:len(input_circuits)]` to exclude the results of the measurement circuits.

ddddddanni replied:
Sure! Since 
```
shuffled_measurements = [res[0] for res in results] -> which is a list of [ResultDict]
```
I also changed the method header to ```tuple[list[ResultDict], SingleQubitReadoutCalibrationResult]```

eliottrosenberg commented on cirq-core/cirq/contrib/shuffle_circuits/shuffle_circuits_with_readout_benchmarking.py:
```diff
@@ -0,0 +1,194 @@
+    random_bitstrings: np.ndarray,
+    readout_repetitions: int,
+    qubits: list["cirq.Qid"],
+) -> dict["cirq.Qid", tuple[float, float]]:
```
Actually, since we already have a convenient class to store this information, `cirq.experiments.single_qubit_readout_calibration.SingleQubitReadoutCalibrationResult`, which has some plotting methods, maybe we can return that instead (and the same for the function from which this is called; it can return a `SingleQubitReadoutCalibrationResult`.

ddddddanni replied:
Done.

eliottrosenberg reviewed (changes requested):
Thanks, Danni! This looks great! Just some small nits.

NoureldinYosri commented on cirq-core/cirq/contrib/shuffle_circuits/shuffle_circuits_with_readout_benchmarking.py:
```diff
@@ -0,0 +1,218 @@
+                input_circuits are used.
+
+    Returns:
+        The unshuffled measurements of input circuits and a dictionary from qubits
```
this seems to describe only one of the returns

ddddddanni replied:
Sorry for the confusion here. Changed the description.

NoureldinYosri commented on cirq-core/cirq/contrib/shuffle_circuits/shuffle_circuits_with_readout_benchmarking.py:
```diff
@@ -0,0 +1,218 @@
+    qubits: list["cirq.Qid"],
+    timestamp: float,
+) -> SingleQubitReadoutCalibrationResult:
+    """Analyzes the readout error rates from the unshuffled measurements."""
```
docstrings for local methods are optional, however this method seems complex enough to deserve a full docstirng

ddddddanni replied:
Sure!

NoureldinYosri reviewed (changes requested):
LGTM%nits

NoureldinYosri commented on cirq-core/cirq/contrib/shuffle_circuits/shuffle_circuits_with_readout_benchmarking.py:
```diff
@@ -0,0 +1,241 @@
+
+    # Run the shuffled circuits and measure
+    results = sampler.run_batch(shuffled_circuits, repetitions=all_repetitions)
+    timestamp = time.time()
```
is this a debug statement?
```suggestion
```

ddddddanni replied:
This is part of the input to _analyze_readout_results method. To construct SingleQubitReadoutCalibrationResult, we need to have the timestamp recording when the data was taken.

NoureldinYosri commented on cirq-core/cirq/contrib/shuffle_circuits/shuffle_circuits_with_readout_benchmarking.py:
```diff
@@ -0,0 +1,241 @@
+            sample_counts = np.sum(trial_result, axis=0)
+
+            zero_state_trials += sample_counts * \
+                (1 - random_bitstrings[trial_idx])
```
so $0 \to 1$ and $1 \to 0$ ... can we change the logic so that we just do `sample_counts  & random_bitstrings[trial_idx]`

---

this more of a style taste so it is up to you, but can you reshape the `random_bitstrings` so that instead of doing `random_bitstrings [trial_idx]` we just iterate with `zip(measurement_result.measurements.items(), bits)`

ddddddanni replied:
Sure, added the zip(measurement_result.measurements.items(), bits)

NoureldinYosri reviewed (commented):
just a few nits

NoureldinYosri reviewed (commented):
@ddddddanni LGTM, please fix the typing and lint issues and this PR will be good to merge

ddddddanni commented:
> @ddddddanni LGTM, please fix the typing and lint issues and this PR will be good to merge
Sure! Fixed the type and format check

eliottrosenberg reviewed (approved):
Great work, @ddddddanni!
